k8s/upgrade/src/plugin/upgrade.rs

```diff
@@ -103,25 +104,80 @@ pub struct UpgradeArgs {
     /// Upgrade to an unsupported version.
     #[clap(global = true, hide = true, long, default_value_t = false)]
     pub skip_upgrade_path_validation_for_unsupported_version: bool,
+
+    /// The set values specified by the user for upgrade.
+    /// (can specify multiple or separate values with commas: key1=val1,key2=val2).
+    #[clap(global = true, long)]
+    pub set: Vec<String>,
+}
+
+impl Default for UpgradeArgs {
+    fn default() -> Self {
+        Self::new()
+    }
 }
 
 impl UpgradeArgs {
+    /// Initialise with default values.
+    pub fn new() -> Self {
+        Self {
+            dry_run: false,
+            skip_data_plane_restart: false,
+            skip_single_replica_volume_validation: false,
+            skip_replica_rebuild: false,
+            skip_cordoned_node_validation: false,
+            skip_upgrade_path_validation_for_unsupported_version: false,
+            set: Default::default(),
+        }
+    }
     ///  Upgrade the resources.
-    pub async fn apply(&self, namespace: &str) {
+    pub async fn apply(&self, namespace: &str) -> error::Result<()> {
+        let gur = UpgradeEventClient::new(namespace).await?;
+        let release_name = get_release_name(namespace).await?;
+
+        // Delete any previous upgrade events
+        gur.delete_upgrade_events(release_name.clone()).await?;
         // Create resources for upgrade
-        _ = UpgradeResources::create_upgrade_resources(
-            namespace,
-            self.skip_data_plane_restart,
-            self.skip_upgrade_path_validation_for_unsupported_version,
-        )
-        .await
-        .map_err(|error| {
-            std::process::exit(error.into());
-        });
-        console_logger::info(UPGRADE_JOB_STARTED, "");
+        UpgradeResources::create_upgrade_resources(namespace, self).await?;
+
+        for _i in 0 .. MAX_RETRY_ATTEMPTS {
+            // wait for 10 seconds for the upgrade event to be published
+            tokio::time::sleep(Duration::from_secs(10)).await;
+            match gur.get_upgrade_events(release_name.clone(), false).await {
+                Ok(latest_event) => self.handle_upgrade_event(latest_event, namespace).await?,
+                Err(_) => continue,
+            }
+            break;
+        }
+
+        Ok(())
+    }
+
+    /// Handle the event and errors out for invalid helm command.
+    async fn handle_upgrade_event(
+        &self,
+        latest_event: Event,
+        namespace: &str,
+    ) -> error::Result<()> {
+        if let Some(action) = latest_event.action {
+            if action == "Failed" {
+                if let Some(data) = latest_event.message {
```

Comment:
I'm confused, this is checking the latest event, how do we know if this is the validation event?